### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by Rocky14683**
#### Summary
Make changes to differential class

1. Added this-> pointer for every class instance
2. Update class instance linearSettings, angularSettings, drivetrain to a shared_ptr
3. Make tank, arcade, and curvature functions take const reference as its parameter
4. Make the Drivetrain constructor take the const reference of shared_ptr as parameters(because you wouldn't want to pass the ownership)
5. Make the code of the makeMotorGroup function cleaner and make it take const reference as parameters

## Download the template for this pull request: 

> [!NOTE]  
> This is auto generated from [`Add Template to Pull Request`](https://github.com/LemLib/LemLib/actions/runs/6987336329)
- via manual download: [LemLib@0.5.0-rc1-cd0b79a.zip](https://nightly.link/LemLib/LemLib/actions/artifacts/1072556314.zip)
- via PROS Integrated Terminal: 
 ```
curl -o LemLib@0.5.0-rc1-cd0b79a.zip https://nightly.link/LemLib/LemLib/actions/artifacts/1072556314.zip;
pros c fetch LemLib@0.5.0-rc1-cd0b79a.zip;
pros c apply LemLib@0.5.0-rc1-cd0b79a;
rm LemLib@0.5.0-rc1-cd0b79a.zip;
```

### CODE REVIEW

File: include/lemlib/chassis/differential.hpp [hunk 1]
```diff
@@ -294,16 +298,16 @@ class Differential : public Chassis {
...
          *
          */
         void update() override;
 
-        ControllerSettings linearSettings;
-        ControllerSettings angularSettings;
-        Drivetrain drivetrain;
+        std::unique_ptr<ControllerSettings> linearSettings;
+        std::unique_ptr<ControllerSettings> angularSettings;
+        std::unique_ptr<Drivetrain> drivetrain;
```
Comment (liam-teale) [inline include/lemlib/chassis/differential.hpp:311]: Why do these need to be pointers?

File: src/lemlib/chassis/differential.cpp [hunk 2]
```diff
@@ -72,12 +71,12 @@ Differential::Differential(Drivetrain drivetrain, ControllerSettings linearSetti
  */
 void Differential::initialize() {
     // calibrate odom
-    odom->calibrate();
+    this->odom->calibrate();
```
Comment (liam-teale) [inline src/lemlib/chassis/differential.cpp:74]: why are we using `this->` when accessing class members?

**Review by liam-teale — COMMENTED** [review]
Just some questions about some of the design choices on this PR

File: include/lemlib/chassis/differential.hpp (hunk 1 above)
Comment (Rocky14683) [reply include/lemlib/chassis/differential.hpp:311]: these are the advantages:
1. once differential objects got destroyed, they went out of scope and then destroyed themselves
2. compare . vs ->, -> is easier to read(this is one of an reason why ppl use pointer)
3. the differential will have the ownership of these instances

File: src/lemlib/chassis/differential.cpp (hunk 2 above)
Comment (Rocky14683) [reply src/lemlib/chassis/differential.cpp:74]: You should always use this-> for class members to differentiate things from in/outside of the class. It helps ppl to understand the code and allows you to have the function parameters named the same as the class members.

**Comment by liam-teale** [comment]
Since we are switching to always using `this->`, can you make its usage consistent across the codebase?

**Pull request merged**